### PR TITLE
Add split_queries to query stats

### PR DESCRIPTION
**Pull request by ssncferreira**
#### What this PR does

Add split_queries parameter describing the number of partial queries executed in parallel to both range and instant queries.

#### Which issue(s) this PR fixes or relates to

Fixes #<issue number>

#### Checklist

- [ ] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/frontend/querymiddleware/split_and_cache.go [hunk 1]
```diff
@@ -126,6 +127,10 @@ func (s *splitAndCacheMiddleware) Do(ctx context.Context, req Request) (Response
 		return nil, err
 	}
 
+	// Update query stats.
+	queryStats := stats.FromContext(ctx)
+	queryStats.AddSplitQueries(uint32(len(splitReqs)))
```
Comment (pracucci) [inline pkg/frontend/querymiddleware/split_and_cache.go:132]: I think it's better to check the actual number of downstream requests after the cache has been looked up. Basically, I would report `execReqs`, which is the actual number of split queries we'll run.

**Review by pracucci — APPROVED** [review]
Thanks! Overall LGTM, but look at my comment, please.

Can we also add an assertion to 1 on the existing tests to check it's actually added to the query stats?

**Comment by ssncferreira** [comment]
> Can we also add an assertion to 1 on the existing tests to check it's actually added to the query stats?

@pracucci added tests for stats: https://github.com/grafana/mimir/pull/2641/commits/34835d527214a3f4bdac3610a9d92fabf6f3518e

However, I don't see any tests that assert the written log line. Let me know if I'm missing something, otherwise I can create a unit test to assert this 🙂

**Comment by ssncferreira** [comment]
Noticed that the ruler also logs query's stats: https://github.com/grafana/mimir/blob/main/pkg/ruler/compat.go#L195

Since the split by range feature also affects the ruler, maybe it is also worthwhile adding it here 🤔

**Comment by pracucci** [comment]
> Noticed that the ruler also logs query's stats: https://github.com/grafana/mimir/blob/main/pkg/ruler/compat.go#L195
> 
> Since the split by range feature also affects the ruler, maybe it is also worthwhile adding it here 🤔

The ruler is not affected by the splitting. Splitting is done only by query-frontend. The ruler has two evaluations modes:
- Internal: no splitting, requests don't go through query-frontend
- Remote: requests go though query-frontend, there's splitting supported but provided by the query-frontend so it will be logged by the query-frontend middleware

**Comment by ssncferreira** [comment]
> > Noticed that the ruler also logs query's stats: https://github.com/grafana/mimir/blob/main/pkg/ruler/compat.go#L195
> > Since the split by range feature also affects the ruler, maybe it is also worthwhile adding it here 🤔
> 
> The ruler is not affected by the splitting. Splitting is done only by query-frontend. The ruler has two evaluations modes:
> 
> * Internal: no splitting, requests don't go through query-frontend
> * Remote: requests go though query-frontend, there's splitting supported but provided by the query-frontend so it will be logged by the query-frontend middleware

Ah, yes I meant the ruler will receive the split requests from the query-frontend (remote point). But already understood that that will be logged in the query-frontend. Thanks for the explanation 🙂

**Pull request merged**